src/contract/client.ts

```diff
@@ -12,20 +12,79 @@ import { processSpecEntryStream } from './utils';
 const CONSTRUCTOR_FUNC = "__constructor";
 
 async function specFromWasm(wasm: Buffer) {
-  const wasmModule = await WebAssembly.compile(wasm);
-  const xdrSections = WebAssembly.Module.customSections(
-    wasmModule,
-    "contractspecv0"
-  );
-  if (xdrSections.length === 0) {
+  const customData = parseWasmCustomSections(wasm);
+  const contractSpecData = customData.get('contractspecv0');
+
+  if (!contractSpecData) {
     throw new Error("Could not obtain contract spec from wasm");
   }
-  const bufferSection = Buffer.from(xdrSections[0]);
+
+  const bufferSection = Buffer.from(contractSpecData[0]);
   const specEntryArray = processSpecEntryStream(bufferSection);
   const spec = new Spec(specEntryArray);
+
   return spec;
 }
 
+function parseWasmCustomSections(buffer: Buffer): Map<string, Uint8Array[]> {
+  const sections = new Map<string, Uint8Array[]>();
+  const arrayBuffer = buffer.buffer.slice(
+    buffer.byteOffset,
+    buffer.byteOffset + buffer.byteLength
+  );
+
+  let offset = 0;
+
+  // Helper to read bytes with bounds checking
+  const read = (length: number): Uint8Array => {
+    if (offset + length > buffer.byteLength) throw new Error('Buffer overflow');
+    const bytes = new Uint8Array(arrayBuffer, offset, length);
+    offset += length;
+    return bytes;
+  };
+
+  // Validate header
+  if ([...read(4)].join() !== '0,97,115,109') throw new Error('Invalid WASM magic');
+  if ([...read(4)].join() !== '1,0,0,0') throw new Error('Invalid WASM version');
+
+  while (offset < buffer.byteLength) {
+    const sectionId = read(1)[0];
+    const sectionLength = readVarUint32();
+    const start = offset;
+
+    if (sectionId === 0) { // Custom section
+      const nameLen = readVarUint32();
+
+      if (nameLen === 0 || offset + nameLen > start + sectionLength) continue;
+
+      const nameBytes = read(nameLen);
+      const payload = read(sectionLength - (offset - start));
+
+      try {
+        const name = new TextDecoder('utf-8', { fatal: true }).decode(nameBytes);
+        if (payload.length > 0) {
+          sections.set(name, (sections.get(name) || []).concat(payload));
+        }
+      } catch { /* Invalid UTF-8 */ }
```

Comment:
Shouldn't this re-throw?

Reply:
Personally I don't think so. We really only care about fully valid custom sections and ultimately only the `contractspecv0` so rather than deal with edge cases we just skip over them.